### PR TITLE
redo redpanda container lifecycle probes

### LINKED ISSUE #421: awk for startupProbe and readinessProbe only look for controller_id in down nodes, not current broker id
It appears that when doing the startupProbe and readinessProbe for pod health, we aren't actually looking for the current pod's broker id. We're looking for the controller id in the down nodes list, which could trigger a cascading failure of actually healthy nodes. Additionally, we also may be missing cases where a broker _is_ in the list, but we don't fail out the probe. It doesn't appear that we have a good way to find the current broker id in the list based on the health_overview.

https://github.com/redpanda-data/helm-charts/blob/6e31833a0df989a76229a4f12522bbff3d107aff/charts/redpanda/templates/statefulset.yaml#L216-L247

See thread for more detail:

https://redpandadata.slack.com/archives/C01H6JRQX1S/p1680276099505199

From the thread:

Assume we have 3 brokers in a cluster. controller is on broker 0. If broker id 0 shows up in the down nodes list, all pods in the cluster will likely die off.

Additionally, if any other broker shows up in the down nodes list, the awk ends up with a false negative and doesn't seem to exit with a non-zero status to trigger a failure in the probes for that broker.

I instrumented the awk a bit to make this more obvious in testing it. Below we can see that broker0 being down in the first test would trigger the probe failure. But if broker1 was down, we don't trigger.

```
$ cat health1
{"is_healthy": true, "controller_id": 0, "all_nodes": [0 1 2], "nodes_down": [0], "leaderless_partitions": []}
$
$ bash awk health1
id start: {"is_healthy": true, "controller_id": 0, "all_nodes": [0 1 2], "nodes_down": [0], "leaderless_partitions": []}
id first gsub: 0, "all_nodes": [0 1 2], "nodes_down": [0], "leaderless_partitions": []}
id second gsub: 0
nd_str start: {"is_healthy": true, "controller_id": 0, "all_nodes": [0 1 2], "nodes_down": [0], "leaderless_partitions": []}
nd_str first gsub: 0], "leaderless_partitions": []}
nd_str second gsub: 0
nodes_down: nd_list 1 items - nd_list[i] 0 nodes_down[nd_list[i]]  end
(exit value) id in nodes_down:  1
$
$ cat health3
{"is_healthy": true, "controller_id": 0, "all_nodes": [0 1 2], "nodes_down": [1], "leaderless_partitions": []}
$
$ bash awk health3
id start: {"is_healthy": true, "controller_id": 0, "all_nodes": [0 1 2], "nodes_down": [1], "leaderless_partitions": []}
id first gsub: 0, "all_nodes": [0 1 2], "nodes_down": [1], "leaderless_partitions": []}
id second gsub: 0
nd_str start: {"is_healthy": true, "controller_id": 0, "all_nodes": [0 1 2], "nodes_down": [1], "leaderless_partitions": []}
nd_str first gsub: 1], "leaderless_partitions": []}
nd_str second gsub: 1
nodes_down: nd_list 1 items - nd_list[i] 1 nodes_down[nd_list[i]]  end
(exit value) id in nodes_down:  0
```

### PR DESCRIPTION
**Pull request by joejulian**
Fixes #421

### CODE REVIEW

File: charts/redpanda/Chart.yaml [hunk 1]
```diff
@@ -23,7 +23,7 @@ type: application
 # The chart version and the app version are not the same and will not track
 # together. The chart version is a semver representation of changes to this
 # chart.
-version: 4.0.0
+version: 4.0.1
```
Comment (mmaslankaprv) [inline charts/redpanda/Chart.yaml:26]: nit: this could be a separate commit

**Review by RafalKorepta — APPROVED** [review]

File: charts/redpanda/templates/statefulset.yaml [hunk 2]
```diff
@@ -215,83 +215,38 @@ spec:
...
                 - -c
-                {{- if (include "admin-internal-tls-enabled" . |fromJson).bool }}
-                - >
-                  curl https://localhost:{{ .Values.listeners.admin.port }}/v1/cluster/health_overview
-                  -svk --cacert /etc/tls/certs/{{ .Values.listeners.admin.tls.cert }}/ca.crt
-                {{- else }}
-                - >
-                  curl -sv http://localhost:{{ .Values.listeners.admin.port }}/v1/cluster/health_overview
-                {{- end }}
+                - curl --silent --fail -k {{ include "admin-tls-curl-flags" . }} "{{ include "admin-http-protocol" . }}://localhost:{{ .Values.listeners.admin.port }}/v1/node_config"
```
Comment (jcsp) [inline charts/redpanda/templates/statefulset.yaml:234]: If you use `/v1/ready`, that's a bit cheaper for the server than serializing the node config each time.

You don't necessarily have to read the body of the response ("ready" or "booting"), if you just want a simple endpoint to check if the admin API is up.

File: charts/redpanda/templates/statefulset.yaml [hunk 3]
```diff
@@ -215,83 +215,38 @@ spec:
...
-                  awk '{
-                    id = $0; gsub(/.*"controller_id": /, "", id); gsub(/,.*/, "", id)
-                    nd_str = $0; gsub(/.*"nodes_down": \[/, "", nd_str); gsub(/\].*/, "", nd_str)
-                    FS=","
-                    split(nd_str, nd_list)
-                    for (i in nd_list) nodes_down[nd_list[i]]=""
-                    exit (id in nodes_down)
-                  }'
-                {{- end }}
+                - curl --silent --fail -k {{ include "admin-tls-curl-flags" . }} "{{ include "admin-http-protocol" . }}://localhost:{{ .Values.listeners.admin.port }}/v1/node_config" | grep -o '"node_id":[0-9]\+'
```
Comment (jcsp) [inline charts/redpanda/templates/statefulset.yaml:224]: Perhaps this check should be followed by a call to `/v1/ready` to check the status is `ready`.

As-written, this startupProbe will complete slightly before the node is actually ready to serve Kafka requests.

The startup order is:
- Bootstrap (including getting a node ID if necessary)
- Start most other services (including the admin server, which will return "booting" from /v1/ready at this point)
- Start listening on the kafka port
- Finally, set the status to "ready" in the result of /v1/ready.
- Bootstrap (this includes getting a node ID if it doesn't already have one)

Comment (alejandroEsc) [reply charts/redpanda/templates/statefulset.yaml:224]: @jcsp would it be enough just to wait for the /v1/ready to both respond and give an http 200 response?

Comment (joejulian) [reply charts/redpanda/templates/statefulset.yaml:224]: Looks like a "no". It gives 200 for "booting" which is not "ready".

File: charts/redpanda/templates/statefulset.yaml (hunk 2 above)
Comment (joejulian) [reply charts/redpanda/templates/statefulset.yaml:234]: Thanks!

File: charts/redpanda/Chart.yaml (hunk 1 above)
Comment (joejulian) [reply charts/redpanda/Chart.yaml:26]: We like to keep them together. It's handy for `git log -S`.

File: charts/redpanda/templates/statefulset.yaml (hunk 3 above)
Comment (joejulian) [reply charts/redpanda/templates/statefulset.yaml]: I just switched to the ready link. That seems to be sufficient for the startupProbe

Comment (joejulian) [reply charts/redpanda/templates/statefulset.yaml]: had to use `/v1/status/ready` as `/v1/ready` was giving 404s.

**Review by alejandroEsc — APPROVED** [review]
this is an awesome improvement.
